test/server/lib/DocApi.ts

```diff
@@ -842,6 +842,218 @@ function testDocApi() {
     assert.equal(resp.status, 200);
   });
 
+  describe("PUT /docs/{did}/columns", function () {
+
+    async function generateDocAndUrl() {
+      const wid = (await userApi.getOrgWorkspaces('current')).find((w) => w.name === 'Private')!.id;
+      const docId = await userApi.newDoc({name: 'ColumnsPut'}, wid);
+      const url = `${serverUrl}/api/docs/${docId}/tables/Table1/columns`;
+      return { url, docId };
+    }
+
+    async function getColumnFieldsMapById(url: string) {
+      const result = await axios.get(url, chimpy);
+      assert.equal(result.status, 200);
+      return new Map(
+          result.data.columns.map(
+            ({id, fields}: {id: string, fields: object}) => [id, fields]
+          )
+      );
+    }
+
+    const COLUMN_TO_ADD = {
+      id: "Foo",
+      fields: {
+        type: "Text",
+        label: "FooLabel",
+      }
+    };
+
+    const EXISTING_COLUMN_LABEL = "A";
+    const COLUMN_TO_UPDATE = {
+      id: "A",
+      fields: {
+        type: "Numeric",
+        colId: "NewA"
+      }
+    };
+
+    it('should create new columns', async function () {
+
+      // given
+      const { url } = await generateDocAndUrl();
+      const body: ColumnsPut = {
+        columns: [COLUMN_TO_ADD]
+      };
+      // when
+      const resp = await axios.put(url, body, chimpy);
+      // then
+      const columnFieldsMap = await getColumnFieldsMapById(url);
+      assert.equal(resp.status, 200);
+
+      assert.isTrue(columnFieldsMap.has(COLUMN_TO_ADD.id), `Expecting to have a column with id "${COLUMN_TO_ADD.id}"`);
+      assert.deepInclude(columnFieldsMap.get(COLUMN_TO_ADD.id), COLUMN_TO_ADD.fields);
+    });
+
+    it('should update existing columns and create new ones', async function () {
+      // given
+      const { url } = await generateDocAndUrl();
+      const submittedColumns: ColumnsPut = {
+        columns: [COLUMN_TO_ADD, COLUMN_TO_UPDATE]
+      };
+
+      // when
+      const resp = await axios.put(url, submittedColumns, chimpy);
+
+      // then
+      assert.equal(resp.status, 200);
+      const fieldsByColId = await getColumnFieldsMapById(url);
+
+      const updatedColFields = fieldsByColId.get(COLUMN_TO_UPDATE.fields.colId);
```

Comment:
I find these assertions difficult to read and not helpful. I also don't like the duplicated logic. Please refactor into something like this:

```
async function checkPut(
  columns: RecordWithStringId[],
  params: Record<string, any>,
  expectedFieldsByColId: Record<string, Record<string, any>>,
) {
  const {url} = await generateDocAndUrl();
  const body: ColumnsPut = {columns};
  const resp = await axios.put(url, body, {...chimpy, params});
  assert.equal(resp.status, 200);
  const fieldsByColId = await getColumnFieldsMapById(url);
  assert.deepEqual([...fieldsByColId.keys()], Object.keys(expectedFieldsByColId));
  for (const [colId, expectedFields] of Object.entries(expectedFieldsByColId)) {
    assert.include(fieldsByColId[colId], expectedFields);
  }
}

await checkPut([COLUMN_TO_ADD, COLUMN_TO_UPDATE], {}, {
  NewA: {type: "Numeric", label: "A"}, B: {}, C: {}, Foo: COLUMN_TO_ADD.fields,
});
```

Reply:
Indeed, that's much better, thanks for the suggestion!

I also changed a little bit the implementation. I don't attempt to get the hidden columns, because otherwise I delete this column:
```js
  {
    id: 'manualSort',
    fields: {
      colRef: 1,
      parentId: 1,
      parentPos: 1,
      type: 'ManualSortPos',
      widgetOptions: '',
      isFormula: false,
      formula: '',
      label: 'manualSort',
      description: '',
      untieColIdFromLabel: false,
      summarySourceCol: 0,
      displayCol: 0,
      visibleCol: 0,
      rules: null,
      recalcWhen: 0,
      recalcDeps: null
    }
  }
```

 However, I am not sure at all of what I am doing:
 - I don't know what is the function of this column (is this for ordering the rows manually as its name suggests?);
 - I don't know whether `includeHidden` may also fetch columns hidden manually by the user (but I guess these columns are just hidden from a view, not from a raw table, right?);

Thanks in advance!

Reply:
> I don't know what is the function of this column (is this for ordering the rows manually as its name suggests?)

Yes

> I don't know whether includeHidden may also fetch columns hidden manually by the user (but I guess these columns are just hidden from a view, not from a raw table, right?);

Right, they're not hidden from a raw table. This is for special hidden helper columns like display columns. If the synchronizing code isn't aware of them, things could go bad. I suggest either implementing allowing `/columns` to return hidden columns, or use `/tables/_grist_Tables_column/records` instead.

Reply:
> Right, they're not hidden from a raw table. This is for special hidden helper columns like display columns. If the synchronizing code isn't aware of them, things could go bad. I suggest either implementing allowing /columns to return hidden columns, or use /tables/_grist_Tables_column/records instead.

I see. I have also updated the GET API to allow including the hidden columns: bda7abfc6a48b97e0c23153686a92a42e0cd392f

Also I don't let `replaceall` remove the hidden columns, I would rather let a `DELETE` request do that manually (assuming they are not much of them).